raidtools2-0.90.20010914/raidstart.8
.\" -*- nroff -*-
.TH raidstart 8
.SH NAME
raidstart, raidstop, \- command set to manage md devices.
.SH SYNOPSIS

.BI raidstart
[options] <raiddevice>*

.BI raidstop
[options] <raiddevice>

.SH DESCRIPTION
RAID devices are virtual devices created from two or more real block
devices. This allows multiple disks to be combined into a single
filesystem, possibly with automated backup and recovery. Linux RAID
devices are implemented through the md device driver.

If you're using the
.B /proc
filesystem,
.B /proc/mdstat
gives you informations about md devices status.

Currently, Linux supports linear md devices, RAID0 (striping), 
RAID1 (mirrroring), and RAID4 and RAID5. For information on
the various levels of RAID, check out:

     http://ostenfeld.dk/~jakob/Software-RAID.HOWTO/

And for instruction for upgrading from a regular system to software raid:

     http://www.tldp.org/HOWTO/Boot+Root+Raid+LILO.html

For new releases of the RAID driver check out:

     ftp://ftp.fi.kernel.org/pub/linux/daemons/raid/alpha

Avaible commands are :

.br
.B mkraid
: configures (creates) md (RAID) devices in the kernel, banding multiple devices into one.
.br

.B raidstart
: activates (starts) an existing 'persistent' md device
.br

.B raid0run
: activates old nonpersistent RAID0/LINEAR md devices
.br

.B raidstop
: turns off an md device, and unconfigures (stops) it
.br

By default, a systems RAID configuration is kept in \fB/etc/raidtab\fR, which
can configure multiple RAID devices. 

All of these tools work similiarly. If \fB-a\fR (or \fB-\-all\fR) is
specified, the specified operation is performed on all of the RAID
devices mentioned in the configuration file. Otherwise, one or more
RAID devices must be specified on the command line. For example:

.B raid0run -a

Starts all of the 'old' RAID0 RAID devices specified in \fB/etc/raidtab\fR. If
only \fI/dev/md1\fR should be started, the following command
should be used instead:

.B raidstart /dev/md1


.SH OPTIONS
.IP "\fB\-a\fP, \fB\-\-all\fP"
Apply the command to all of the configurations specified in the config file.

.IP "\fB-c\fP, \fB\-\-configfile\fP \fIfilename\fR"
Use \fIfilename\fR as the configuration file (\fB/etc/raidtab\fR is used
by default).

.IP "\fB\-h\fP, \fB\-\-help\fP"
Displays a short usage message, then exits.

.IP "\fB\-V\fP, \fB\-\-version\fP"
Displays a short version message, then exits.

.SH NOTES

The raidtools are derived from the md-tools and raidtools packages, which
were originally written by Marc Zyngier, Miguel de Icaza, Gadi Oxman, 
Bradley Ward Allen, and Ingo Molnar.

.SH BUGS
no known bugs.

.SH SEE ALSO
.IR raidtab (5),
.IR raid0run (8),
.IR raidstop (8),
.IR mkraid (8)


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



raidtools2-0.90.20010914/raidstart.c

/*
   raidstart.c : Multiple Devices tools for Linux
               Copyright (C) 1994-96 Marc ZYNGIER
	       <zyngier@ufr-info-p7.ibp.fr> or
	       <maz@gloups.fdn.fr>
	       Copyright (C) 1998 Erik Troan
	       <ewt@redhat.com>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.
   
   You should have received a copy of the GNU General Public License
   (for example /usr/src/linux/COPYING); if not, write to the Free
   Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.  
*/

#include "common.h"
#include "parser.h"
#include "raidlib.h"
#include "version.h"

#include <popt.h>
#include <string.h>
#include <asm/page.h>
#include <sys/sysmacros.h>

#define RAIDADD_VERSION "v0.3d"


void usage(unsigned char *arg) {
    printf("usage: %s [--all] [--configfile] [--help] [--version] [-achv] </dev/md?>\n", arg);
}

int main (int argc, char *argv[]) {
    int i, all=0, readonly=0, readwrite=0, exit_status=0, version = 0, help = 0;
    char *namestart=argv[0];
    md_cfg_entry_t *p;
    char ** args;
    enum raidFunc func;
    FILE * fp;
    poptContext optCon;
    char * configFile = RAID_CONFIG;
    struct poptOption optionsTable[] = {
	{ "all", 'a', 0, &all, 0 },
	{ "configfile", 'c', POPT_ARG_STRING, &configFile, 0 },
	{ "readonly", 'r', 0, &readonly, 0 },
	{ "readwrite", 'w', 0, &readwrite, 0 },
	{ "help", 'h', 0, &help, 0 },
	{ "version", 'V', 0, &version, 0 },
	{ NULL, 0, 0, NULL, 0 }
    } ;

    namestart = strrchr(argv[0], '/');
    if (!namestart)
	namestart = argv[0];
    else
	namestart++;

    if (!strcmp (namestart, "raidstart"))
	func = raidstart;
    else if (!strcmp (namestart, "raidstop"))
	func = raidstop;
    else if (!strcmp (namestart, "raidhotadd"))
	func = raidhotadd;
    else if (!strcmp (namestart, "raidhotremove"))
	func = raidhotremove;
    else if (!strcmp (namestart, "raidsetfaulty"))
	func = raidsetfaulty;
    else if (!strcmp (namestart, "raidhotgenerateerror"))
	func = raidhotgenerateerror;
    else {
	fprintf (stderr, "Unknown command %s\n", argv[0]);
	usage(namestart);
	return (EXIT_FAILURE);
    }

    optCon = poptGetContext(namestart, argc, (const char **)argv,
			    optionsTable, 0);
    if ((i = poptGetNextOpt(optCon)) < -1) {
	fprintf(stderr, "%s: %s\n", 
		poptBadOption(optCon, POPT_BADOPTION_NOALIAS),
		poptStrerror(i));
	usage(namestart);
	return EXIT_FAILURE;
    }

    if (help) {
	usage(namestart);
	return(EXIT_USAGE);
    } else if (version) {
	printf ("%s " RAIDADD_VERSION " compiled for md " MD_VERSION "\n",
		argv[0]);
	return (EXIT_VERSION);
    }

    fp = fopen(configFile, "r");
    if (fp == NULL) {
        fprintf(stderr, "Couldn't open %s -- %s\n", configFile, strerror(errno));
        exit(EXIT_FAILURE);
    }

    if (prepare_raidlib())
         return EXIT_FAILURE;

    if (readwrite) {
        int fd;
        if (func != raidstart) {
	   fprintf(stderr, "Can do --readwrite only with raidstart!\n");
	   return EXIT_FAILURE;
        }
	args = (char **)poptGetArgs(optCon);
	if (!args) {
	    fprintf(stderr, "nothing to do!\n");
            usage(namestart);
	    exit(EXIT_FAILURE);
	}

	fd = open_or_die(*args);

        if (do_raidstart_rw (fd, *args))
	    exit(EXIT_FAILURE);
        exit(0);
    }

    if (readonly) {
         if (func != raidstop) {
	    fprintf(stderr, "Can do --readonly only with raidstop!\n");
	    return EXIT_FAILURE;
         }
         func = raidstop_ro;
    }

    /*
     * stop is special, we want to get it done
     * without parsing the config
     */
    if ((func == raidstop) || (func == raidstop_ro)) {
        int fd;

	args = (char **)poptGetArgs(optCon);
	if (!args) {
	    fprintf(stderr, "nothing to do!\n");
            usage(namestart);
	    exit(EXIT_FAILURE);
	}

	fd = open_or_die(*args);

        if (do_raidstop (fd, *args, func == raidstop ? 0:1))
	    exit(EXIT_FAILURE);
        exit(0);
    }

    /*
     * we want to fast-handle hot-remove and hot-add too.
     */
    if ((func == raidhotremove) || (func == raidhotadd) ||
			 (func == raidhotgenerateerror) || (func == raidsetfaulty)) {
        int md_fd;
	char * md_name;

	md_name = (char *)poptGetArg(optCon);
	if (!md_name) {
	    fprintf(stderr, "nothing to do!\n");
            usage(namestart);
	    exit(EXIT_FAILURE);
	}

	
	md_fd = open_or_die(md_name);

	args = (char **)poptGetArgs(optCon);
	if (!args) {
	    fprintf(stderr, "no target disk specified!\n");
            usage(namestart);
	    exit(EXIT_FAILURE);
	}

	if (func == raidhotremove) {
	        if (do_raidhotremove (md_fd, *args, md_name))
		    exit(EXIT_FAILURE);
	} else {
		if (func == raidhotadd) {
		        if (do_raidhotadd (md_fd, *args, md_name))
			    exit(EXIT_FAILURE);
		} else if (func == raidhotgenerateerror) {
		        if (do_raidhotgenerateerror (md_fd, *args, md_name))
			    exit(EXIT_FAILURE);
		} else if (func == raidsetfaulty) {
			    if (do_raidsetfaulty (md_fd, *args, md_name))
			    exit(EXIT_FAILURE);
		}
	}
        exit(0);
    }


    if (parse_config(fp))
	exit(EXIT_FAILURE);

    if (all) {
	for (p = cfg_head; p; p = p->next)
	    exit_status += handleOneConfig(func, p);
    } else {
	args = (char **)poptGetArgs(optCon);
	if (!args) {
	    fprintf(stderr, "nothing to do!\n");
            usage(namestart);
	    return EXIT_FAILURE;
	}

	while (*args) {
	    for (p = cfg_head; p; p = p->next) {
		if (strcmp(p->md_name, *args)) continue;
		exit_status += handleOneConfig(func, p);
		break;
	    }
	    if (!p) {
		fprintf(stderr, "device %s is not described in config file\n", *args);
		exit_status++;
	    }
	    args++;
	}
    }

    poptFreeContext(optCon);

    return (exit_status);
}

